KOGITO-8303: [SWF Dev UI] Add support for starting workflows using gateway API and cloud events (#291)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/tooling/quarkus-dev-ui-extension/quarkus-dev-ui-workflow-definition-page.adoc

= Workflow Definitions
:compat-mode!:
// Metadata:
:description: Workflow Definitions
:keywords: kogito, workflow, serverless, Quarkus, Dev UI, definitions

> In Kogito Serverless Workflow Tools extension, the *Workflow Definitions* page displays the list of available workflows in

Reviewer: ```suggestion
In the Kogito Serverless Workflow Tools extension, the *Workflow Definitions* page displays the list of available workflows in
```

> the Kogito project and helps you to start new workflow instances either via REST or by triggering Cloud Events.

Reviewer: ```suggestion
the Kogito project and helps you to start new workflow instances either via REST or by triggering cloud events.
```

.Workflow Definitions page
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-workflow-definitions.png[]

> In *Workflow Definitions* table you'll find the list of workflows that shows the following details:

Reviewer: ```suggestion
In the *Workflow Definitions* table you will find the list of workflows that shows the following details:
```

* *Workflow Name*: Displays the name of a workflow definition.
* *Endpoint*: Displays the URL to a workflow definition.
* *Actions*: Provides the *Start new Workflow* button to start workflow instances via REST.

> In the table toolbar provides the ability to apply filters to the table or to trigger cloud events that can start workflow instances

Reviewer: ```suggestion
In the table, toolbar provides the ability to apply filters to the table or to trigger cloud events that can start workflow instances
```

> == Applying filters to the Workflow Definitions table

Reviewer: Is this a procedure module? If yes, then we can convert the following sentences into numbered steps.

> To search for a specific workflow definition, enter the name of the workflow in the *Filter by workflow name* field and

Reviewer: Something like this?
```suggestion
. To search for a specific workflow definition, enter the workflow's name in the *Filter by workflow name* field.
. Click the *Apply Filter* button. The matching workflow definitions appear as a chip below the search field.
. Click the *Reset to default* button to clear the applied filters. 
. To fetch newly added workflow definitions, click on the refresh icon next to the *Apply Filter* button.
```

[[con-start-workflow-instances]]
== Starting new Workflow instances
Workflows can be started either by using the specific workflow REST endpoints or by triggering Cloud Events.
The Kogito Serverless Workflow Tools extension allows you to use both mechanisms.

[[con-start-workflow-instances-rest]]
=== Starting new Workflow instances using REST
If you want to start a new workflow instance using the workflow REST endpoint, just click on the *Start new Workflow*
button of any of the workflow in the *Workflow Definitions* table, then you'll be redirected to the *Start New Workflow*
page where you could setup the data and Business Key that will be used to start the new workflow instance.

=== Filling up the Workflow data
Depending on your workflow configuration the page can provide different mechanisms to fill the workflow data.
If your workflow is configured with a valid _Data Input Schema_, this page will be able to display a custom Form to help
a filling the data.

.Start Workflow Page with a Form defined in the _Data Input Schema_
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-form.png[]

> If your workflow doesn't provide a _Data Input Schema_, the page will display a Code Editor that will enable you to

Reviewer: ```suggestion
If your workflow does not provide a _Data Input Schema_, the page will display a Code Editor that will enable you to manually fill the workflow data in JSON format.
```

manually fill the workflow data in JSON format.

> .Start Workflow Page with a  Code Editor.

Reviewer: ```suggestion
.Start Workflow Page with a Code Editor.
```

=== Starting the new Workflow instance
Clicking on the *Start* button will POST the workflow data and the Business Key to the workflow REST endpoint. If the
workflow instance starts successfully, a success alert appears on the top of the screen, which contains the
*Go to workflow list* link to navigate to the xref:tooling/quarkus-dev-ui-extension/quarkus-dev-ui-workflow-instances-page.adoc[Workflow Instances page].

.Example of workflow successful starting notification
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-success-alert.png[]

If there is an issue while starting a workflow, then a failure alert appears on the top of the screen, containing *View Details* and *Go to workflow list* options. The *View Details* enables you to view the error message.

.Example of workflow starting failure notification
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-fail-alert.png[]

[[con-trigger-cloud-event-page]]
== Starting Workflows using Cloud Events
On the *Workflow Definitions* Page, the *Trigger Cloud Event* button enables you to start new workflow instances by using

> HTTP Cloud Events. This will forward you to the *Trigger Cloud Event* page.

Reviewer: ```suggestion
HTTP Cloud Events. This will lead you to the *Trigger Cloud Event* page.
```

> Once there, you'll have to fill the form with the Cloud Event information:

Reviewer: ```suggestion
Once there, you will have to fill out the form with the Cloud Event information:
```

- *Endpoint*: Defines the Path and the HTTP method where the cloud event should be triggered.
- *Event Type*: Sets the Cloud Event type header.
- *Event Source*: Sets the Cloud Event Source header. Defaults to `/local/quarkus-devUi`.
- *Business Key*: Sets the new workflow instance Business Key, it will be added in the `kogitobusinesskey` event header.
- *Event Custom Headers*: Allows you to define custom headers to be added to the event.

> - *Event Data*: data that will be in the event in JSON format.

Reviewer: ```suggestion
- *Event Data*: Data that will be in the event in JSON format.
```

.Start Workflow Page with a Code Editor.
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-no-form.png[]

[NOTE]
====
For more information about how to setup the Input Schema Definition on your {context}, please take a look at the
xref:core/defining-an-input-schema-for-workflows.adoc[Input Schema for {context}] section.
====

=== Setting up the Business Key
You can use the *Business key* text box to define a custom business key value to the workflow instance.
If the *Business Key* field is blank, then an auto-generated business key is defined for the workflow instance.

=== Starting the new Workflow instance
By clicking on the *Start* button will POST the workflow data and the Business Key to the workflow REST endpoint. If the
workflow instance starts successfully, a success alert appears on the top of the screen, which contains the
*Go to workflow list* link to navigate to the xref:tooling/quarkus-dev-ui-extension/quarkus-dev-ui-workflow-instances-page.adoc[Workflow Instances page].

.Example of workflow successful starting notification
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-success-alert.png[]

> If there is an issue while starting a workflow, then a failure alert appears on the top of the screen, containing the*View Details* and *Go to workflow list* options. The *View Details* enables you to view the error message.

Reviewer 2: ```suggestion
If there is an issue while starting a workflow, then a failure alert appears on the top of the screen, containing the *View Details* and *Go to workflow list* options. The *View Details* enables you to view the error message.
```

the Kogito project and helps you to start new workflow instances either via REST or by triggering cloud events.

.Workflow Definitions page
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-workflow-definitions.png[]

In the *Workflow Definitions* table you will find the list of workflows that shows the following details:

* *Workflow Name*: Displays the name of a workflow definition.
* *Endpoint*: Displays the URL to a workflow definition.
* *Actions*: Provides the *Start new Workflow* button to start workflow instances via REST.

In the table, the toolbar provides the ability to apply filters to the table or to trigger cloud events that can start workflow instances

== Applying filters to the Workflow Definitions table
. To search for a specific workflow definition, enter the workflow's name in the *Filter by workflow name* field.
. Click the *Apply Filter* button. The matching workflow definitions appear as a chip below the search field.
. Click the *Reset to default* button to clear the applied filters. 
. To fetch newly added workflow definitions, click on the refresh icon next to the *Apply Filter* button.

> click *Apply Filter* button. The matching workflow definitions appear as a chip below the search field.

Reviewer: ```suggestion
Click the *Apply Filter* button. The matching workflow definitions appear as a chip below the search field.
```

button of any of the workflow in the *Workflow Definitions* table, then you'll be redirected to the *Start New Workflow*
page where you could setup the data and Business Key that will be used to start the new workflow instance.

=== Filling up the Workflow data
Depending on your workflow configuration the page can provide different mechanisms to fill the workflow data.
If your workflow is configured with a valid _Data Input Schema_, this page will be able to display a custom Form to help
a filling the data.

.Start Workflow Page with a Form defined in the _Data Input Schema_
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-form.png[]

If your workflow does not provide a _Data Input Schema_, the page will display a Code Editor that will enable you to manually fill the workflow data in JSON format.
manually fill the workflow data in JSON format.

.Start Workflow Page with a Code Editor.
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-no-form.png[]

[NOTE]
====
For more information about how to setup the Input Schema Definition on your {context}, please take a look at the

> xref:core/defining-an-input-schema-for-workflows.adoc[Input Schema for {context}] section..

Reviewer 2: ```suggestion
xref:core/defining-an-input-schema-for-workflows.adoc[Input Schema for {context}] section.
```

====

=== Setting up the Business Key

> You can use the *Business key* text box to define a custom business key value to the workflow instance.

Reviewer 2: maybe give more details about what business key is what for what it is used for?

If the *Business Key* field is blank, then an auto-generated business key is defined for the workflow instance.

=== Starting the new Workflow instance

> Clicking on the *Start* button will POST the workflow data and the Business Key to the workflow REST endpoint. If the

Reviewer 2: ```suggestion
Clicking on the *Start* button will send a POST request with the workflow data and the Business Key to the workflow REST endpoint. If the
```

workflow instance starts successfully, a success alert appears on the top of the screen, which contains the

> *Go to workflow list* link to navigate to the xref:tooling/quarkus-dev-ui-extension/quarkus-dev-ui-workflow-instances-page.adoc[Workflow Instances page].

Reviewer 2: left over?

.Example of workflow successful starting notification
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-success-alert.png[]

> If there is an issue while starting a workflow, then a failure alert appears on the top of the screen, containing *View Details* and *Go to workflow list* options. The *View Details* enables you to view the error message.

Reviewer 2: ```suggestion
If there is an issue while starting a workflow, then a failure alert appears on the top of the screen, containing the*View Details* and *Go to workflow list* options. The *View Details* enables you to view the error message.
```

.Example of workflow starting failure notification
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-start-workflow-fail-alert.png[]

[[con-trigger-cloud-event-page]]
== Starting Workflows using Cloud Events
On the *Workflow Definitions* Page, the *Trigger Cloud Event* button enables you to start new workflow instances by using
HTTP Cloud Events. This will lead you to the *Trigger Cloud Event* page.

Once there, you will have to fill out the form with the Cloud Event information:

- *Endpoint*: Defines the Path and the HTTP method where the cloud event should be triggered.
- *Event Type*: Sets the Cloud Event type header.
- *Event Source*: Sets the Cloud Event Source header. Defaults to `/local/quarkus-devUi`.
- *Business Key*: Sets the new workflow instance Business Key, it will be added in the `kogitobusinesskey` event header.
- *Event Custom Headers*: Allows you to define custom headers to be added to the event.
- *Event Data*: Data that will be in the event in JSON format.

.Starting a workflow using a cloud event
image::tooling/quarkus-dev-ui-extension/kogito-swf-tools-trigger-cloud-events.png[]

> Click on the *Trigger* button to trigger the Cloud Event. If the workflow instance starts successfully, a success alert appears on the top of the screen, which contains the

Reviewer: ```suggestion
Click the *Trigger* button to trigger the cloud event. If the workflow instance starts successfully, a success alert appears on the top of the screen, which contains the
```
